### ISSUE TITLE
Subdomains for PEPCO and Delmarva Power are incorrect for some states

### ISSUE THREAD

**Issue by trancefam** [tronikos/opower#14]
Hi,

I was trying to setup the Home Assistant plugin and kept getting this error (not sure if related or separate issue):

```yml
2023-08-02 22:02:58.615 ERROR (MainThread) [homeassistant.components.opower.coordinator] Unexpected error fetching Opower data: 'NoneType' object has no attribute 'group'
Traceback (most recent call last):
File "/lsiopy/lib/python3.11/site-packages/homeassistant/helpers/update_coordinator.py", line 283, in _async_refresh
self.data = await self._async_update_data()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/lsiopy/lib/python3.11/site-packages/homeassistant/components/opower/coordinator.py", line 67, in _async_update_data
await self.api.async_login()
File "/lsiopy/lib/python3.11/site-packages/opower/opower.py", line 151, in async_login
self.access_token = await self.utility.async_login(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/lsiopy/lib/python3.11/site-packages/opower/utilities/exelon.py", line 44, in async_login
settings = json.loads(re.search(r"var SETTINGS = ({.*});", result).group(1))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'group'
```

So instead I tried to use this python package directly outside of HA, which gave me the error:

```
raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 403, message='Forbidden', url=URL('https://pep.opower.com/ei/edge/apis/multi-account-v1/cws/pep/customers?offset=0&batchSize=100&addressFilter=')
```

When loading https://secure.pepco.com/api/Services/MyAccountService.svc/GetConfiguration I can see the following values

![image](https://github.com/tronikos/opower/assets/26947075/d1f59593-4940-4cd6-abe7-14e981d549f2)

When logging into my account, all requests seem to prefer `oPowerURLBaseJurisdiction`:

![image](https://github.com/tronikos/opower/assets/26947075/3e37ef2c-fa19-4873-b9a5-13f0a8775a22)

I was able to get the python script working by [changing this](https://github.com/tronikos/opower/blob/8d7f2a99e313c07a1975b3a50f40b2c950ac7d84/src/opower/utilities/pepco.py#L18C16-L18C21) to `pepd`.

I am unsure if this is dynamic, as it seemed to work when it was initially added in #3

**Comment by tronikos** [comment]
Interesting. It seems for some jurisdictions it's pep while for others it's pepd. 

Evergy is another utility that has two different domains depending on location, see https://github.com/tronikos/opower/pull/9 and https://github.com/tronikos/opower/pull/11

You need to figure out how the JS code picks the correct domain and replicate the logic in Python.

**Comment by ronaldheft** [comment]
I can take a look later today and see if I can make sense of this. I currently live in MD and I used to live in DC, so I have an old DC account that should be helpful to test the theory.

**Comment by tronikos** [comment]
Delmarva Power (states of Delaware and Maryland) is another utility that seems to use two opower subdomains depending on state.

https://secure.delmarva.com/api/Services/MyAccountService.svc/GetConfiguration returns:
dpl.opower.com and dpld.opower.com

So your solution should ideally work for both Pepco and Delmarva Power.

**Comment by ronaldheft** [comment]
I haven't found a great solution yet. My Pepco account has a mix of MD and DC addresses. My MD address uses `pep.opower.com`. My DC address uses `pepd.opower.com`.

Using Charles, I can see both the `oPowerURLBase` and `oPowerURLBaseJurisdiction` keys remain the same (`pep` and `pepd`, respectively) regardless of address. If I search for `oPowerURLBaseJurisdiction`, I don't see any uses of the key in JavaScript or headers.

The best solution I can think of would be to call `https://secure.pepco.com/.euapi/mobile/custom/auth/accounts` and look at `data.[].PremiseInfo.[].MainAddress.townDetail.stateOfProvince`. If `state == "DC"` use `pepd`? Feels fragile and there has to be another spot in the code that determines which endpoint is called.

**Comment by trancefam** [comment]
@ronaldheft I found usage of `oPowerURLBaseJurisdiction` in https://secure.pepco.com/dist/app.js?ver=633315 :

```js
var opowerUrl = Exelon.Web.Configuration.OpCoConfigurationObject.OpowerURLBase;
'dpl' !== $scope.opco && 'pepco' !== $scope.opco || !$scope.isResidential || 'MD' !== $scope.state &&
(
  opowerUrl = Exelon.Web.Configuration.OpCoConfigurationObject.OpowerURLBaseJurisdiction
);
```

I was able to find this by loading this page https://secure.pepco.com/MyAccount/MyBillUsage/Pages/Secure/ViewMyUsage.aspx

After formatting `app.js` it is here:
![image](https://github.com/tronikos/opower/assets/26947075/dce3a8c0-6e3f-498e-8b11-f879b4903d86)

Perhaps we could implement similar logic, or maybe a easier approach is just always try `oPowerURLBase` and if it fails, try `oPowerURLBaseJurisdiction`?

**Comment by tronikos** [comment]
Great find. BGE also loads https://secure.bge.com/dist/app.js?ver=633315 that has the exact same code:
```js
var opowerUrl = Exelon.Web.Configuration.OpCoConfigurationObject.OpowerURLBase;
'dpl' !== $scope.opco &&
'pepco' !== $scope.opco ||
!$scope.isResidential ||
'MD' !== $scope.state &&
(
  opowerUrl = Exelon.Web.Configuration.OpCoConfigurationObject.OpowerURLBaseJurisdiction
);
```
It's just a matter of finding which endpoint returns the state then to replicate this logic in Python.

**Comment by trancefam** [comment]
Just breaking down the source of the arguments:
- `opco`
  - comes from https://secure.pepco.com/api/Services/MyAccountService.svc/GetConfiguration ![image](https://github.com/tronikos/opower/assets/26947075/3769043e-2694-4a07-a898-65b604e36964)
- `isResidential`
  - comes from https://secure.pepco.com/.euapi/mobile/custom/auth/accounts at `data[i].isResidential` ![image](https://github.com/tronikos/opower/assets/26947075/f532091c-75d5-482e-b66d-52773cb8d299)
- `state`
  - comes from https://secure.pepco.com/.euapi/mobile/custom/auth/accounts at `data[i].PremiseInfo[0].mainAddress.townDetail.stateOrProvince`  ![image](https://github.com/tronikos/opower/assets/26947075/10b378e8-cbfe-4874-b7f8-7c072bfca416)

One thing to be mindful of for both `isResidential` and `state`, the values can be different for each address on the account, which could result in different oPower URL to use.

**Comment by tronikos** [comment]
Possible solutions:
1) Get `isResidential` and `state` from `/.euapi/mobile/custom/auth/accounts`
2) Add new subclasses for PEPCO and Delmarva Power that just have a different name and subdomain
3) Have some fallback mechanism and use the `oPowerURLBaseJurisdiction` if `oPowerURLBase` fails.

I'll let you decide which solution you implement. My preference is 1 or if that's too hard do 2.

**Comment by trancefam** [comment]
@tronikos I implemented option 1 in PR #23
